AI/Nullkiller/Engine/PriorityEvaluator.cpp

```diff
@@ -273,7 +299,8 @@ float RewardEvaluator::getEnemyHeroStrategicalValue(const CGHeroInstance * enemy
 		vstd::amax(objectValue, getStrategicalValue(obj));
 	}
 
-	return objectValue / 2.0f + enemy->level / 15.0f;
+	//AI should absolutely prioritize killing enemy heroes, even scouts
+	return std::min(1.0f, objectValue * 0.9f + (1.0f - (1.0f / (1 + enemy->level))));
```

Comment:
I think 1.0 it is too high. Scout should not be higher than capturing town with fort in current turn. And better add some basic value than use min(1, value). so that 2 lvl hero is more valuable than 1 lvl. Also we can increase this radius getOneTurnAccessibleObjects.
Just a case which I do not want to get broken. When enemy hero approaches a town you buy and put scouts not far from it and hero attacks them loosing movement points. While you bring your main army to protect the town. This is how original AI did afaik.
Mine idea here is simple. AI captures town and all scouts around become a treat for it. Thus they gain additional strategical value equivalent to this town and AI should eliminate them as well. Probably we can express hero value as a separate fuzzy variable to be able to write more flexible rules. If main hero will start pursuing a scout and the last will constantly run away we might end up having main hero just wasting movement points which is not nice as well. Having a separate variable will allow us writing rules like 
if enemy_hero is weak and turn is now then value is bit_high (or high)
if enemy_hero is strong and turn is now then value is highest

Reply:
Well, I tested this in game AI vs AI and it works. AI finally kills the scouts in lategame instead of playing ping-pong, capturing a town and losing it next turn for weeks. Killing 4 scouts in a a turn is a winning strategy and running past them is not.

Reply:
ALso, I think enemy shouldn't know hero level in first place :P

Reply:
AI knows a bit more about you than human unfortunately. It knows exact army structure and your stats. And this is true for both  VCAI and Nullkiller afaik. But I agree that it should not. So this more a problem of client/server architecture that this info is available to AI. If AI knows - it is possible to build modified client which will also show all these stats.
Ok, in current circumstances this might be a better solution. But what I actually wanted was "What you say would make more sense if AI actually tried to buy units in a town to defend it". Still I think 1 is to high. Lets try 0.7f + instead of std:min(1, and see if it gives what you want.

Reply:
But it means that 1.0f is the upper cap and the value can still be lower for lesser objects.

Reply:
I misunderstood it, sorry. Yes this way it looks ok.